src/front/glsl/functions.rs

```diff
@@ -524,6 +525,26 @@ impl Program<'_> {
     }
 }
 
+fn should_read(built_in: BuiltIn, stage: ShaderStage) -> bool {
```

Comment:
This looks like a list that the validator uses. Let the validator take care of this. Here, if you need to exclude specific things, then make it an exclusion list specifically, instead of listing all the possibilities.

Reply:
The problem here is that the validator complains if a builtin is added as a function argument in a stage where it shouldn't

Reply:
I don't quite understand. If you have an entry point in GLSL that's using some built-in, then just add it to the arguments. If it's not using this built-in, and you aren't adding it to the arguments, then the validator have nothing to complain about.

Reply:
You have the case where the built in is being used but shouldn't be written for example (vertex stage)
```glsl
void main() {
  gl_Postition = vec4(1);
}
```
without this fix the backend would produce this pseudocode
```glsl
vec4 position;

void wrapper(
  [builtin=position] vec4 x
) -> [builtin=position] vec4 {
  position = x;
  main();
  return position;
}

void main() {
  position = vec4(1);
}
```
this is invalid because position can't be read in the vertex stage, with this change the generated pseudocode would look like this

```glsl
vec4 position;

void wrapper() -> [builtin=position] vec4 {
  main();
  return position;
}

void main() {
  position = vec4(1);
}
```

Reply:
Thank you for writing this down! I think there is a problem in the first step. Why does the pseudocode have "position = x" at all? When GLSL is parsed, we should see that `gl_Position` (not specifically it, anything goes) is on the left side of an assignment, so there is no reading going on.

Reply:
The frontend doesn't keep global use information even though it should because with multiple stages we can't blindly emit globals arguments, I think we should add this for now and I'll make another PR which collects the global use of the functions and the functions it calls recursively and only emit the loads and the stores for the return based on it.